Support glm3 and glm4.

youth123 opened the pull request:
I have fixed the issues mentioned in https://github.com/ggerganov/llama.cpp/pull/6999. This code can totally supports glm3 and glm4 model architecture and can be emdded in ollama server. This PR is based on https://github.com/mnlife/llama.cpp/tree/glm4 and https://github.com/mnlife/llama.cpp/tree/chatglm3, by @mnlife and @xingxingqiao.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

xingxingqiao committed:
remove .rotary_pos_emb.inv_freq and unuse code for chatglm3 model

Signed-off-by: XingXing Qiao <qiaoxx@dingdao.com>

xingxingqiao committed:
optimize convert-hf-to-gguf.py for chatglm model

Signed-off-by: XingXing Qiao <qiaoxx@dingdao.com>

xingxingqiao committed:
fix lint error

Signed-off-by: XingXing Qiao <qiaoxx@dingdao.com>

xingxingqiao committed:
add chatglm3-6b model support huggingface model:
 https://hf-mirror.com/THUDM/chatglm3-6b

Signed-off-by: XingXing Qiao <qiaoxx@dingdao.com>

xingxingqiao committed:
support glm-4-9b-chat

Signed-off-by: XingXing Qiao <qiaoxx@dingdao.com>

youth123 committed:
fix eos tokens to glm4

youth123 committed:
remove unused log

youth123 committed:
Fix eos tokens to glm4 and adapts to glm3

matteoserva commented on llama.cpp:
```diff
@@ -18386,8 +18599,18 @@ int32_t llama_tokenize(
+        if (LLAMA_VOCAB_TYPE_SPM == vocab_type) {
+            prompt = "[gMASK]sop<|user|>\n" + prompt + "<|assistant|>";
+        } else if (LLAMA_VOCAB_TYPE_BPE == vocab_type) { // glm4
+            prompt = "[gMASK]<sop><|user|>\n" + prompt + "<|assistant|>";
```
The tokenizer shouldn't modify the prompt unless add_special is specified. The current behavious breaks the /tokenizer endpoint in the server

youth123 replied:
Ok, I will put this part of the code inside the tokenizer when determine the vocab type.

arch-btw commented:
This is so great! Thank you 👍 !

There are only a couple of things that I ran into:

During compile, a small note:

```
llama.cpp: In function ‘int32_t llama_tokenize(const llama_model*, const char*, int32_t, llama_token*, int32_t, bool, bool)’:
llama.cpp:18603:28: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
18603 |     auto prompt = std::move(std::string(text, text_len));
      |                   ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
llama.cpp:18603:28: note: remove ‘std::move’ call

```

And `convert-hf-to-gguf.py` doesn't work with bf16. 

But it does work with f32.

bf16 log:

```
INFO:hf-to-gguf:output.weight,             torch.bfloat16 --> BF16, shape = {4096, 151552}
Writing:   0%|                                                                                                                                                                     | 0.00/18.8G [00:00<?, ?byte/s]Traceback (most recent call last):
  File "/home/glm4/convert-hf-to-gguf.py", line 3072, in <module>
    main()
  File "/home/glm4/convert-hf-to-gguf.py", line 3066, in main
    model_instance.write()
  File "/home/glm4/convert-hf-to-gguf.py", line 331, in write
    self.gguf_writer.write_tensors_to_file(progress=True)
  File "/home/glm4/gguf-py/gguf/gguf_writer.py", line 312, in write_tensors_to_file
    ti.tensor.tofile(self.fout)
  File "/home/glm4/gguf-py/gguf/lazy.py", line 233, in tofile
    eager = LazyNumpyTensor.to_eager(self)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/glm4/gguf-py/gguf/lazy.py", line 193, in to_eager
    return cls._recurse_apply(t, simple_to_eager)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/glm4/gguf-py/gguf/lazy.py", line 109, in _recurse_apply
    return fn(o)
           ^^^^^
  File "/home/glm4/gguf-py/gguf/lazy.py", line 185, in simple_to_eager
    lt._data = lt._func(lt._args)
               ^^^^^^^^^^^^^^^^^^
  File "/home/glm4/gguf-py/gguf/lazy.py", line 158, in <lambda>
    return cls(meta=cls.eager_to_meta(res), lazy=shared_lazy, args=args, func=lambda a: fn(*a, **kwargs))
                                                                                        ^^^^^^^^^^^^^^^^
  File "/home/glm4/gguf-py/gguf/quants.py", line 52, in __quantize_bf16_array
    return __apply_over_grouped_rows(__compute_fp32_to_bf16, arr=n, otype=np.int16, oshape=n.shape)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/glm4/gguf-py/gguf/quants.py", line 47, in __apply_over_grouped_rows
    np.concatenate([func(group).ravel() for group in np.array_split(rows, n_groups)], axis=0, out=out)
                    ^^^^^^^^^^^
  File "/home/glm4/gguf-py/gguf/quants.py", line 30, in __compute_fp32_to_bf16
    n = np.where((n & 0x7fffffff) > 0x7f800000, (n & 0xffff0000) | (64 << 16), n)
                                                 ~~^~~~~~~~~~~~
OverflowError: Python integer 4294901760 out of bounds for int32
```

Other than that it works great.

**Prompt:**

> ./llama-cli -m glm4-Q6_K.gguf --color -p "[gMASK]<sop><|system|>\nYou are a helpful assistant<|user|>\nHello<|assistant|>\n"

**Output:**

```
You are a helpful assistant
Hello

Hello! How can I assist you today? [end of text]
```

youth123 commented:
> This is so great! Thank you 👍 !
> 
> There are only a couple of things that I ran into:
> 
> During compile, a small note:
> 
> ```
> llama.cpp: In function ‘int32_t llama_tokenize(const llama_model*, const char*, int32_t, llama_token*, int32_t, bool, bool)’:
> llama.cpp:18603:28: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
> 18603 |     auto prompt = std::move(std::string(text, text_len));
>       |                   ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> llama.cpp:18603:28: note: remove ‘std::move’ call
> ```
> 
> And `convert-hf-to-gguf.py` doesn't work with bf16.
> 
> But it does work with f32.
> 
> bf16 log:
> 
> ```
> INFO:hf-to-gguf:output.weight,             torch.bfloat16 --> BF16, shape = {4096, 151552}
> Writing:   0%|                                                                                                                                                                     | 0.00/18.8G [00:00<?, ?byte/s]Traceback (most recent call last):
>   File "/home/glm4/convert-hf-to-gguf.py", line 3072, in <module>
>     main()
>   File "/home/glm4/convert-hf-to-gguf.py", line 3066, in main
>     model_instance.write()
>   File "/home/glm4/convert-hf-to-gguf.py", line 331, in write
>     self.gguf_writer.write_tensors_to_file(progress=True)
>   File "/home/glm4/gguf-py/gguf/gguf_writer.py", line 312, in write_tensors_to_file
>     ti.tensor.tofile(self.fout)
>   File "/home/glm4/gguf-py/gguf/lazy.py", line 233, in tofile
>     eager = LazyNumpyTensor.to_eager(self)
>             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/glm4/gguf-py/gguf/lazy.py", line 193, in to_eager
>     return cls._recurse_apply(t, simple_to_eager)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/glm4/gguf-py/gguf/lazy.py", line 109, in _recurse_apply
>     return fn(o)
>            ^^^^^
>   File "/home/glm4/gguf-py/gguf/lazy.py", line 185, in simple_to_eager
>     lt._data = lt._func(lt._args)
>                ^^^^^^^^^^^^^^^^^^
>   File "/home/glm4/gguf-py/gguf/lazy.py", line 158, in <lambda>
>     return cls(meta=cls.eager_to_meta(res), lazy=shared_lazy, args=args, func=lambda a: fn(*a, **kwargs))
>                                                                                         ^^^^^^^^^^^^^^^^
>   File "/home/glm4/gguf-py/gguf/quants.py", line 52, in __quantize_bf16_array
>     return __apply_over_grouped_rows(__compute_fp32_to_bf16, arr=n, otype=np.int16, oshape=n.shape)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/glm4/gguf-py/gguf/quants.py", line 47, in __apply_over_grouped_rows
>     np.concatenate([func(group).ravel() for group in np.array_split(rows, n_groups)], axis=0, out=out)
>                     ^^^^^^^^^^^
>   File "/home/glm4/gguf-py/gguf/quants.py", line 30, in __compute_fp32_to_bf16
>     n = np.where((n & 0x7fffffff) > 0x7f800000, (n & 0xffff0000) | (64 << 16), n)
>                                                  ~~^~~~~~~~~~~~
> OverflowError: Python integer 4294901760 out of bounds for int32
> ```
> 
> Other than that it works great.
> 
> **Prompt:**
> 
> > ./llama-cli -m glm4-Q6_K.gguf --color -p "[gMASK]<|system|>\nYou are a helpful assistant<|user|>\nHello<|assistant|>\n"
> 
> **Output:**
> 
> ```
> You are a helpful assistant
> Hello
> 
> Hello! How can I assist you today? [end of text]
> ```

I'll go take a look at this issue.

youth123 committed:
add preprocess to chatglm3 and chatglm4

arch-btw commented:
Thanks, I think it's related to my pip environment and not a problem with the code.

ngxson commented on llama.cpp:
```diff
@@ -18324,6 +18550,19 @@ llama_token_attr llama_token_get_attr(const struct llama_model * model, llama_to
 bool llama_token_is_eog(const struct llama_model * model, llama_token token) {
+    auto arch_name = llama_model_arch_name(model->arch);
+    auto vocab_type = model->vocab.type;
+    if (strcmp(arch_name, "chatglm") == 0) {
```
`llama_token_is_eog` is called quite often, doing string compare here may have impact on performance

ngxson replied:
Looking at [tokenizer_config.json](https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/tokenizer_config.json), personally I think that it's safe to stop at EOS (`<|endoftext|>`), so no need to hard-code token IDs here

ngxson replied:
Edit: looking at chat template, seems like the model does not have the notion end-of-turn token (strange!). Maybe we need to introduce EOT token as a list instead of single value. This will require adding metadata to gguf (CC @ggerganov )

youth123 replied:
Alright, I will add an eot list to the metadata of gguf. Then, during the initialization of vocab, I will put all the eot entries into this variable. At that time, the judgment will only require traversing this list.

youth123 replied:
I have already added the eod_id_list variable to the gguf meta and ensured compatibility with the previous versions. Could you please check if there are any other modifications needed?

https://github.com/ggerganov/llama.cpp/pull/8031/files#diff-4f653096980bd7d10518aa909cb648452cd3aa380ff93cb9fb642dca48536526R110

https://github.com/ggerganov/llama.cpp/pull/8031/files#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR5090

ngxson commented on llama.cpp:
```diff
@@ -18714,6 +18952,28 @@ static int32_t llama_chat_apply_template_internal(
+        if (add_ass) {
+            ss << "<|assistant|>";
+        }
+    } else if (tmpl == "ChatGLM4") {
```
Naming should be consistent (i.e. all low-cap like `chatgml3`)

ngxson replied:
```suggestion
    } else if (tmpl == "chatgml4" || tmpl.find("[gMASK]<sop>") != std::string::npos) {
```

ngxson commented on llama.cpp:
```diff
@@ -18714,6 +18952,28 @@ static int32_t llama_chat_apply_template_internal(
+    } else if (tmpl == "chatglm3" ||
+        (tmpl.find("add_generation_prompt") != std::string::npos &&
+        tmpl.find("for message in messages") != std::string::npos &&
+        tmpl.find("loop.first") != std::string::npos)) {
```
These conditions are redundant and mostly clash with other chat templates. I'd suggest:

```
tmpl == "chatglm3" || tmpl.find("[gMASK]sop") != std::string::npos
```

Source: https://huggingface.co/THUDM/chatglm3-6b-128k/blob/main/tokenizer_config.json

youth123 committed:
add eos_id_list to llama.cpp

youth123 committed:
fix conflicts

youth123 committed:
fix code style

youth123 committed:
fix code style

youth123 committed:
fix conflicts

youth123 committed:
fix conflicts

youth123 committed:
Merge branch 'glm_support'

ggerganov commented on llama.h:
```diff
@@ -853,7 +856,7 @@ extern "C" {
     // Special tokens
     LLAMA_API llama_token llama_token_bos(const struct llama_model * model); // beginning-of-sentence
-    LLAMA_API llama_token llama_token_eos(const struct llama_model * model); // end-of-sentence
+    LLAMA_API void llama_token_eos(const struct llama_model * model, llama_token* token_list); // end-of-sentence
```
This does not look great. Could you summarize why these models require multiple EOS tokens?

If it is really needed, we might introduce:

```c
LLAMA_API llama_token llama_token_eos_ith(const struct llama_model * model, int i);
```

But first need to understand if this is really necessary

ngxson replied:
@ggerganov Normally, a chat looks like this:

```
<|user|>Who are you<|end|><|assistant|>I'm am AI<|end|>
```

But because this model does not have EOT (end-of-turn) token, the chat looks like this:

```
<|user|>Who are you<|assistant|>I'm am AI<|user|>
```

So the EOT should be `<|user|>`. The list of EOT token contains `<|user|>`, `<|observation|>` and `<|endoftext|>`

But after second thought @youth123 , I realize that can we set `<|endoftext|>` as EOS and `<|user|>` as EOG? We can ignore `<|observation|>` for now, because it's inside generated text (I think?). So finally no need for adding list of EOS (sorry for not thinking about this earlier)

youth123 replied:
Multiple EOS tokens are used to enable the language model to support function calls and tool calls, making it easier to distinguish between roles. I will implement it according to the solution provided by @ngxson . However, for future scalability, I still hope that the official support for multiple EOS tokens will be provided.

ngxson replied:
@youth123 Yeah I'd agree that multiple EOT is very useful for tool calls. However, `llama-cli` and `llama-server` don't yet have support for tool calls. I gave a try with "functionary" a while ago, but I'm not very satisfied with the implementation so it's not get merged. Hopefully we will be able to revisit it in the future.

ggerganov replied:
The `llama_token_is_eog()` was introduced to solve this exact problem:

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/include/llama.h#L847-L850

You need to list all such tokens here:

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/src/llama.cpp#L2339-L2347

For example, add `id special_obs_id = -1;`

In `llm_load_vocab()` initialize from the KV header:

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/src/llama.cpp#L5063-L5078

Or try to match them by name (not recommended):

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/src/llama.cpp#L5107-L5128

Extend `llama_token_is_eog()`:

```c++
bool llama_token_is_eog(const struct llama_model * model, llama_token token) {
    return token != -1 && (
        token == llama_token_eos(model) ||
        token == llama_token_eot(model) ||
        token == llama_token_obs(model)
    );
}
```

youth123 replied:
Okay, thank you. I will follow these instructions to make the changes.

youth123 replied:
I have already set eot and eos in convert-hf-gguf. 
https://github.com/youth123/llama.cpp/blob/0595f03dd171cdad54aafe79199f223d7a2e3f0b/convert-hf-to-gguf.py#L3052C9-L3053C89
``` python
        special_vocab._set_special_token("eos", tokenizer.get_added_vocab()["<|endoftext|>"])
        special_vocab._set_special_token("eot", tokenizer.get_added_vocab()["<|user|>"])
```
Then, when I merged the latest code, I found a bug. There is a risk of type conversion overflow in the llama_chat_apply_template function in common.cpp. When tmpl does not match the current template, res will become negative and then be forcibly converted to uint.
``` c++
std::string llama_chat_apply_template(const struct llama_model * model,
        const std::string & tmpl,
        const std::vector<llama_chat_msg> & msgs,
        bool add_ass) {
    int alloc_size = 0;
    std::vector<llama_chat_message> chat;
    for (auto & msg : msgs) {
        chat.push_back({msg.role.c_str(), msg.content.c_str()});
        alloc_size += (msg.role.size() + msg.content.size()) * 1.25;
    }

    const char * ptr_tmpl = tmpl.empty() ? nullptr : tmpl.c_str();
    std::vector<char> buf(alloc_size);

    // run the first time to get the total output length
    int32_t res = llama_chat_apply_template(model, ptr_tmpl, chat.data(), chat.size(), add_ass, buf.data(), buf.size());

    // if it turns out that our buffer is too small, we resize it
    if ((size_t) res > buf.size()) {
        buf.resize(res);
        res = llama_chat_apply_template(model, ptr_tmpl, chat.data(), chat.size(), add_ass, buf.data(), buf.size());
    }

    std::string formatted_chat(buf.data(), res);
    return formatted_chat;
}
```

ngxson replied:
> Then, when I merged the latest code, I found a bug. There is a risk of type conversion overflow in the llama_chat_apply_template function in common.cpp. When tmpl does not match the current template, res will become negative and then be forcibly converted to uint.

`std::string llama_chat_apply_template()` is a thin wrapper for the underlaying c style `llama_chat_apply_template`, so it's required to use `llama_chat_verify_template` to check if the template is supported or not (if not, we exit). You can see this check in `gpt_params_find_arg`

But I think it's OK to add a bit redundant check in the cpp function here, can be as simple as

```cpp
if (res < 0) {
  throw std::runtime_error("template is not supported");
}
```

youth123 replied:
Ok, I understand. However, it seems that the input parameters must include chat-template; otherwise, this check will not be effective.

youth123 committed:
Revert "add eos_id_list to llama.cpp"

This reverts commit 3a4d5790bfdc205c5b658204239f168fc21cc1a8.

youth123 commented:
> Thank you again for your work. I noticed that the output ends in an unexpected way.
> 
> Example call: `./llama-cli -t 4 -m ../glm-4-9b-chat-Q6_K.gguf -s 0 --temp 0 -e -p "[gMASK]<sop><|system|>\nYou are a helpful assistant<|user|>\nOutput the first 10 even natural numbers in a row<|assistant|>\n"`
> 
> Returned output:
> 
> ```
> You are a helpful assistant
> Output the first 10 even natural numbers in a row
> 
> 2, 4, 6, 8, 10, 12, 14, 16 [end of text]
> ```
I will take a look at this issue.

dranger003 commented:
> Thanks, I think it's related to my pip environment and not a problem with the code.

@arch-btw Have you found the issue? I'm running into the same issue for some reason.

youth123 committed:
set <|endoftext|> as eos and <|user|> as eot

youth123 committed:
Merge remote-tracking branch 'offical/master'

youth123 committed:
fix chat template bug

ngxson commented on gguf-py/pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "gguf"
-version = "0.9.0"
+version = "0.9.1"
```
Normally we don't bump version when new arch is added, so I'm not sure if this change is necessary

ngxson commented on gguf-py/gguf/constants.py:
```diff
@@ -87,36 +87,35 @@ class SSM:
-        ADD_BOS              = "tokenizer.ggml.add_bos_token"
-        ADD_EOS              = "tokenizer.ggml.add_eos_token"
-        ADD_PREFIX           = "tokenizer.ggml.add_space_prefix"
+        MODEL            = "tokenizer.ggml.model"
```
I'd prefer revert these whitespace changes

compilade replied:
This looks like it reverts the indentation changes from #8055, which means it's probably due to an oversight when merging the latest `master` into this branch.

ngxson commented on convert-hf-to-gguf.py:
```diff
@@ -2893,6 +2896,191 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+        special_vocab.chat_template = "chatglm4"
+        special_vocab.merges = merges
+        # only add special tokens when they were not already loaded from config.json
+        special_vocab._set_special_token("bos", tokenizer.get_added_vocab()["<|endoftext|>"])
```
It seems like BOS should be `[gMASK]`. We should also set `self.gguf_writer.add_add_bos_token(False)` prevent BOS is added twice in chat template

youth123 replied:
From the hf code, it appears that the bos of glm is composed of [gMask]<sop>. I will first assign [gMask] to bos.

ngxson commented on llama.cpp:
```diff
@@ -14532,12 +14738,19 @@ static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab &
                 }
+                // add prefix to chatglm3
+                if (vocab.type_pre == LLAMA_VOCAB_PRE_TYPE_CHATGLM3) {
+                    output.push_back(64790);
```
It would be nice to have comment to understand what is the corresponded tokens for these numbers

arch-btw commented:
@dranger003 thank you that fixed it for me too! I think it's related to `NumPy 2.0` being released last week.

It's odd that upgrading by itself didn't work. I had to completely remove `numpy` and `torch` and then it worked by using your command. 👍

gabrielpondc commented on convert-hf-to-gguf.py:
Sorry, which env of gguf u used, after the revision the issue still showed 
```
python convert-hf-to-gguf.py ../THUDM/glm-4-9b-chat --outtype q8_0 --outfile glm4.gguf
Traceback (most recent call last):
  File "/2T/Langchain-Ch/glm4cpp/convert-hf-to-gguf.py", line 842, in <module>
    class OrionModel(Model):
  File "/2T/Langchain-Ch/glm4cpp/convert-hf-to-gguf.py", line 843, in OrionModel
    model_arch = gguf.MODEL_ARCH.ORION
  File "/root/anaconda3/envs/chatglmcpp/lib/python3.9/enum.py", line 429, in __getattr__
    raise AttributeError(name) from None
AttributeError: ORION
```

gabrielpondc reviewed (commented):
Sorry, which env of gguf u used, after the revision the issue still showed
```
python convert-hf-to-gguf.py ../THUDM/glm-4-9b-chat --outtype q8_0 --outfile glm4.gguf
Traceback (most recent call last):
  File "/2T/Langchain-Ch/glm4cpp/convert-hf-to-gguf.py", line 842, in <module>
    class OrionModel(Model):
  File "/2T/Langchain-Ch/glm4cpp/convert-hf-to-gguf.py", line 843, in OrionModel
    model_arch = gguf.MODEL_ARCH.ORION
  File "/root/anaconda3/envs/chatglmcpp/lib/python3.9/enum.py", line 429, in __getattr__
    raise AttributeError(name) from None
AttributeError: ORION
```

youth123 committed:
add comment to glm prefix and suffix

youth123 committed:
Merge remote-tracking branch 'offical/master'

youth123 commented:
> Sorry, which env of gguf u used, after the revision the issue still showed
> 
> ```
> python convert-hf-to-gguf.py ../THUDM/glm-4-9b-chat --outtype q8_0 --outfile glm4.gguf
> Traceback (most recent call last):
>   File "/2T/Langchain-Ch/glm4cpp/convert-hf-to-gguf.py", line 842, in <module>
>     class OrionModel(Model):
>   File "/2T/Langchain-Ch/glm4cpp/convert-hf-to-gguf.py", line 843, in OrionModel
>     model_arch = gguf.MODEL_ARCH.ORION
>   File "/root/anaconda3/envs/chatglmcpp/lib/python3.9/enum.py", line 429, in __getattr__
>     raise AttributeError(name) from None
> AttributeError: ORION
> ```
This issue seems to be due to an incorrect architectures attribute in your model's config.json file. You can re-download the model configuration from hf.  
https://huggingface.co/THUDM/glm-4-9b-chat

youth123 committed:
fix conflicts and add rope_ratio & ChatGLMForConditionalGeneration

youth123 committed:
merge master

youth123 committed:
fix chat template bug

ngxson commented on src/llama.cpp:
```diff
@@ -5181,7 +5222,7 @@ static void llm_load_vocab(
             }
         }
-
+        
```
white space should be removed (editorconfig)

ngxson commented on src/llama.cpp:
```diff
@@ -19601,6 +19829,25 @@ static int32_t llama_chat_apply_template_internal(
+        if (add_ass) {
+            ss << "<|assistant|>";
+        }
+    } else if (tmpl.find("chatglm4") != std::string::npos || tmpl.find("[gMASK]<sop>") != std::string::npos) {
```
FYI, on master branch, there is a small change to make the code more readable:

```suggestion
    } else if (tmpl == "chatglm4" || tmpl_contains("[gMASK]<sop>")) {
```

ngxson commented on src/llama.cpp:
```diff
@@ -19601,6 +19829,25 @@ static int32_t llama_chat_apply_template_internal(
         if (add_ass) {
             ss << "<|start_header_id|>assistant<|end_header_id|>\n\n";
         }
+    } else if (tmpl.find("chatglm3") != std::string::npos || tmpl.find("[gMASK]sop") != std::string::npos) {
```
Same here: after merging with latest master, you can do:

```suggestion
    } else if (tmpl == "chatglm3" || tmpl_contains("[gMASK]sop")) {
```

ngxson commented on tests/test-chat-template.cpp:
```diff
@@ -57,7 +57,11 @@ int main(void) {
+        // ChatGLM3
+        "{% for message in messages %}{% if loop.first %}[gMASK]sop<|{{ message['role'] }}|>\n {{ message['content'] }}{% else %}<|{{ message['role'] }}|>\n {{ message['content'] }}{% endif %}{% endfor %}{% if add_generation_prompt %}<|assistant|>{% endif %}",
+        // ChatGLM4
+        "chatglm4",
```
```suggestion
        u8"[gMASK]<sop>{% for item in messages %}{% if item['tools'] is defined %}<|system|>\n你是一个名为 ChatGLM 的人工智能助手。你是基于智谱AI训练的语言模型 GLM-4 模型开发的，你的任务是针对用户的问题和要求提供适当的答复和支持。\n\n# 可用工具{% set tools = item['tools'] %}{% for tool in tools %}{% if tool['type'] == 'function' %}\n\n## {{ tool['function']['name'] }}\n\n{{ tool['function'] | tojson(indent=4) }}\n......{% endif %}{% endfor %}{% endif %}{% if item['content'] %}<|{{ item['role'] }}|>{{ item['metadata'] }}\n{{ item['content'] }}{% endif %}{% endfor %}{% if add_generation_prompt %}<|assistant|>{% endif %}",
```

(The `u8` at the beginning of string indicates that this is UTF-8)

Template is taken from https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/tokenizer_config.json

ngxson reviewed (commented):
I couldn't test this atm, but the code looks good to me. Thanks for the contribution @youth123 

FYI, there is also a fix #8198 that may fix the issue with incorrect that response that @matteoserva pointed out.

Let's also wait for approval from @ggerganov before merging

touale commented:
Thank you again for your work. I found that when the model performs a `perplexity` check it returns `nan`.
```
$ CUDA_VISIBLE_DEVICES=0 ./llama-perplexity -m ./output.guff -f ./wiki.test.raw
main: build = 3283 (bbe1926f)
main: built with cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 for x86_64-linux-gnu
main: seed  = 1719723373
llama_model_loader: loaded meta data with 24 key-value pairs and 283 tensors from ./output.guff (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = chatglm
llama_model_loader: - kv   1:                               general.name str              = glm-4-9b-chat
llama_model_loader: - kv   2:                     chatglm.context_length u32              = 131072
llama_model_loader: - kv   3:                   chatglm.embedding_length u32              = 4096
llama_model_loader: - kv   4:                chatglm.feed_forward_length u32              = 13696
llama_model_loader: - kv   5:                        chatglm.block_count u32              = 40
llama_model_loader: - kv   6:               chatglm.attention.head_count u32              = 32
llama_model_loader: - kv   7:            chatglm.attention.head_count_kv u32              = 2
llama_model_loader: - kv   8:   chatglm.attention.layer_norm_rms_epsilon f32              = 0.000000
llama_model_loader: - kv   9:                          general.file_type u32              = 1
llama_model_loader: - kv  10:               chatglm.rope.dimension_count u32              = 64
llama_model_loader: - kv  11:               tokenizer.ggml.add_bos_token bool             = false
llama_model_loader: - kv  12:                     chatglm.rope.freq_base f32              = 500.000000
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  14:                         tokenizer.ggml.pre str              = chatglm-bpe
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,151552]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,151552]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  17:                      tokenizer.ggml.merges arr[str,151073]  = ["Ġ Ġ", "ĠĠ ĠĠ", "i n", "Ġ t",...
llama_model_loader: - kv  18:            tokenizer.ggml.padding_token_id u32              = 151329
llama_model_loader: - kv  19:                tokenizer.ggml.eos_token_id u32              = 151329
llama_model_loader: - kv  20:                tokenizer.ggml.eot_token_id u32              = 151336
llama_model_loader: - kv  21:            tokenizer.ggml.unknown_token_id u32              = 151329
llama_model_loader: - kv  22:                    tokenizer.chat_template str              = chatglm4
llama_model_loader: - kv  23:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:  121 tensors
llama_model_loader: - type  f16:  162 tensors
merges_keyidx: 17
n_merges: 151073
llm_load_vocab: special tokens cache size = 223
llm_load_vocab: token to piece cache size = 0.9732 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = chatglm
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 151552
llm_load_print_meta: n_merges         = 151073
llm_load_print_meta: n_ctx_train      = 131072
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 2
llm_load_print_meta: n_layer          = 40
llm_load_print_meta: n_rot            = 64
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 16
llm_load_print_meta: n_embd_k_gqa     = 256
llm_load_print_meta: n_embd_v_gqa     = 256
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.6e-07
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 13696
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 500.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 131072
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 9B
llm_load_print_meta: model ftype      = F16
llm_load_print_meta: model params     = 9.40 B
llm_load_print_meta: model size       = 17.51 GiB (16.00 BPW) 
llm_load_print_meta: general.name     = glm-4-9b-chat
llm_load_print_meta: EOS token        = 151329 '<|endoftext|>'
llm_load_print_meta: UNK token        = 151329 '<|endoftext|>'
llm_load_print_meta: PAD token        = 151329 '<|endoftext|>'
llm_load_print_meta: LF token         = 151331 '[gMASK]'
llm_load_print_meta: EOT token        = 151336 '<|user|>'
llm_load_print_meta: max token length = 1024
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA A100-SXM4-80GB, compute capability 8.0, VMM: yes
llm_load_tensors: ggml ctx size =    0.14 MiB
llm_load_tensors: offloading 0 repeating layers to GPU
llm_load_tensors: offloaded 0/41 layers to GPU
llm_load_tensors:        CPU buffer size = 17929.97 MiB
.................................................................................
llama_new_context_with_model: n_ctx      = 2048
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 500.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:  CUDA_Host KV buffer size =    80.00 MiB
llama_new_context_with_model: KV self size  =   80.00 MiB, K (f16):   40.00 MiB, V (f16):   40.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     2.31 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =  1488.00 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    12.01 MiB
llama_new_context_with_model: graph nodes  = 1606
llama_new_context_with_model: graph splits = 364

system_info: n_threads = 52 / 104 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 1 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | 
perplexity: tokenizing the input ..
perplexity: tokenization took 1336.87 ms
perplexity: calculating perplexity over 565 chunks, n_ctx=512, batch_size=2048, n_seq=4
perplexity: 16.21 seconds per pass - ETA 38.13 minutes
[1]8.5085,[2]11.3082,[3]11.8973,[4]12.5024,[5]13.1712,[6]14.4324,[7]15.3985,[8]16.1918,[9]17.7532,
[10]18.5259,[11]18.8466,[12]19.8125,[13]21.1505,[14]20.0578,[15]19.3602,[16]19.1104,[17]nan,[18]nan,[19]nan,
[20]nan,[21]nan,[22]nan,[23]nan,[24]nan,[25]nan,[26]nan,[27]nan,[28]nan,[29]nan,
[30]nan...
```

Also when I run the llama model check, it is normal.
```
./llama-perplexity -m //root/.cache/huggingface/big_models/llama-3-chinese-8b-instruct-v3/ggml-model-f16.gguf -f ./wiki.test.raw
main: build = 3283 (bbe1926f)
main: built with cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 for x86_64-linux-gnu
main: seed  = 1719572900
llama_model_loader: loaded meta data with 22 key-value pairs and 291 tensors from //root/.cache/huggingface/big_models/llama-3-chinese-8b-instruct-v3/ggml-model-f16.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = llama-3-chinese-8b-instruct-v3
llama_model_loader: - kv   2:                          llama.block_count u32              = 32
llama_model_loader: - kv   3:                       llama.context_length u32              = 8192
llama_model_loader: - kv   4:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   6:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   7:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv   8:                       llama.rope.freq_base f32              = 500000.000000
llama_model_loader: - kv   9:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                          general.file_type u32              = 1
llama_model_loader: - kv  11:                           llama.vocab_size u32              = 128256
llama_model_loader: - kv  12:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  14:                         tokenizer.ggml.pre str              = llama-bpe
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,128256]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,128256]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  17:                      tokenizer.ggml.merges arr[str,280147]  = ["Ġ Ġ", "Ġ ĠĠĠ", "ĠĠ ĠĠ", "...
llama_model_loader: - kv  18:                tokenizer.ggml.bos_token_id u32              = 128000
llama_model_loader: - kv  19:                tokenizer.ggml.eos_token_id u32              = 128009
llama_model_loader: - kv  20:                    tokenizer.chat_template str              = {% set loop_messages = messages %}{% ...
llama_model_loader: - kv  21:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:  226 tensors
merges_keyidx: 17
n_merges: 280147
llm_load_vocab: special tokens cache size = 256
llm_load_vocab: token to piece cache size = 0.8000 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 128256
llm_load_print_meta: n_merges         = 280147
llm_load_print_meta: n_ctx_train      = 8192
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 8
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 4
llm_load_print_meta: n_embd_k_gqa     = 1024
llm_load_print_meta: n_embd_v_gqa     = 1024
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 14336
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 500000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 8192
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 8B
llm_load_print_meta: model ftype      = F16
llm_load_print_meta: model params     = 8.03 B
llm_load_print_meta: model size       = 14.96 GiB (16.00 BPW) 
llm_load_print_meta: general.name     = llama-3-chinese-8b-instruct-v3
llm_load_print_meta: BOS token        = 128000 '<|begin_of_text|>'
llm_load_print_meta: EOS token        = 128009 '<|eot_id|>'
llm_load_print_meta: LF token         = 128 'Ä'
llm_load_print_meta: EOT token        = 128009 '<|eot_id|>'
llm_load_print_meta: max token length = 256
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 4 CUDA devices:
  Device 0: NVIDIA A100-SXM4-80GB, compute capability 8.0, VMM: yes
  Device 1: NVIDIA A100-SXM4-80GB, compute capability 8.0, VMM: yes
  Device 2: NVIDIA A100-SXM4-80GB, compute capability 8.0, VMM: yes
  Device 3: NVIDIA A100-SXM4-80GB, compute capability 8.0, VMM: yes
llm_load_tensors: ggml ctx size =    0.14 MiB
llm_load_tensors: offloading 0 repeating layers to GPU
llm_load_tensors: offloaded 0/33 layers to GPU
llm_load_tensors:        CPU buffer size = 15317.02 MiB
.........................................................................................
llama_new_context_with_model: n_ctx      = 2048
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 500000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:  CUDA_Host KV buffer size =   256.00 MiB
llama_new_context_with_model: KV self size  =  256.00 MiB, K (f16):  128.00 MiB, V (f16):  128.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     1.96 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =  1260.50 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    12.01 MiB
llama_new_context_with_model: graph nodes  = 1030
llama_new_context_with_model: graph splits = 356

system_info: n_threads = 52 / 104 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 1 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | 
perplexity: tokenizing the input ..
perplexity: tokenization took 418.448 ms
perplexity: calculating perplexity over 564 chunks, n_ctx=512, batch_size=2048, n_seq=4
perplexity: 16.67 seconds per pass - ETA 39.17 minutes
[1]4.7840,[2]5.8377,[3]6.2029,[4]6.7270,[5]7.0772,[6]7.2855,[7]7.6473,[8]8.2439,[9]8.8646,
[10]9.1446,[11]9.2682,[12]9.3960,[13]9.7859,[14]9.3636,[15]9.3063,[16]9.0332,[17]8.9551,[18]9.0666,[19]8.7837,
[20]8.6513,[21]8.6646,[22]8.3099,[23]8.0029,[24]7.8407,[25]7.6153,[26]7.5095,[27]7.4035,[28]7.3077,[29]7.3900,[30]7.3924....
```

youth123 commented:
> Is there any progress?
I am re-modifying the code according to the review comments and merging it into the latest branch. There have been many changes to llama.cpp recently.

youth123 committed:
fix codestyle

youth123 committed:
merge master

youth123 committed:
fix conflicts

youth123 commented:
@dranger003

I reran the perplexity check for GLM3 and GLM4, but did not encounter the issue you mentioned.
Here are my run commands and model weight links.

```python
python convert-hf-to-gguf.py   /root/.cache/huggingface/hub/models--THUDM--glm-4-9b-chat/snapshots/75792d7ee58a335df6943c5d719cc559b64f8e2a/ --outfile test.gguf
./build/bin/llama-perplexity -m test.gguf  -f /root/tmp/wikitext-2-raw/wiki.test.raw
```
https://huggingface.co/THUDM/glm-4-9b-chat
It seems to be the same reason as the question raised by @dranger003 and @arch-btw.  You can try to reinstall the numpy and torch to solve the question.
>  re-installing torch and numpy using --index-url https://download.pytorch.org/whl/cu121 worked... no idea why though.

youth123 committed:
modified the general name of glm model

ggerganov commented on src/llama.cpp:
```diff
@@ -7382,6 +7449,19 @@ static struct ggml_tensor * llm_build_ffn(
                 cur = ggml_sqr(ctx, cur);
                 cb(cur, "ffn_sqr(relu)", il);
             } break;
+        case LLM_FFN_SWIGLU:
+            {
+                // Project to 4h. If using swiglu double the output width, see https://arxiv.org/pdf/2002.05202.pdf
+                int64_t split_point = cur->ne[0] / 2;
+                struct ggml_tensor * x0 = ggml_cont(ctx, ggml_view_2d(ctx, cur, split_point, cur->ne[1], cur->nb[1], 0));
+                struct ggml_tensor * x1 = ggml_cont(ctx, ggml_view_2d(ctx, cur, split_point, cur->ne[1], cur->nb[1], split_point * ggml_element_size(cur)));
+
+                x0 = ggml_silu(ctx, x0);
+                cb(cur, "ffn_silu", il);
+
+                cur = ggml_mul(ctx, x0, x1);
+                cb(cur, "ffn_mul", il);
+            } break;
```
Instead of introducing this new activation path, you can use `LLM_FFN_SILU` + `LLM_FFN_PAR`. Just need to split the `ffn_up` tensor into `ffn_up` + `ffn_gate` during conversion

youth123 replied:
The order of the split tensor running formulas is reversed. Assume that the split tensors are `x0` and `x1`. The original implementation is `silu(x0) * x1`. Now if `x1` is assigned to gate, the executed formula is `silu(x0 * x1)`. So it should still be necessary to add an implementation of swigrelu.

ggerganov replied:
It's not reversed, `LLM_FFN_PAR` computes exactly `silu(x0) * x1`. But I just noticed that we already have the same code for Phi3, so in that case we can introduce `LLM_FFN_SWIGLU` and reuse it in both models

ggerganov commented on src/llama.cpp:
```diff
@@ -14777,12 +14979,19 @@ static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab &
                     output.push_back(vocab.special_bos_id);
                     is_prev_special = true;
                 }
+                // add prefix to chatglm3
+                if (vocab.type_pre == LLAMA_VOCAB_PRE_TYPE_CHATGLM3) {
+                    output.push_back(64790); // [gMask]
+                    output.push_back(64792); // sop
+                    output.push_back(64795); // <|user|>
+                    output.push_back(30910); // \n
+                    output.push_back(13);
+                }
```
These prefixes and suffixes should not by applied here. Instead, the user code is responsible for that (e.g. chat templates)

ngxson replied:
@youth123 I didn't noticed this part, maybe this explain why the model's answer is incorrect: because the `[gMask]sop<|user|>` is added twice, once here and once in the chat template. We should remove it here.

youth123 replied:
Incorrect answers have nothing to do with being added twice. I have compared the input of the hf model output before, and the output of each layer of the transformer is percentile aligned. But the error becomes larger in the last layer, and I'm still looking at this strange problem now. And regarding chat template, will it run in conversation mode in the future?

Forevery1 commented:
> > Is there any progress?
> 
> I am re-modifying the code according to the review comments and merging it into the latest branch. There have been many changes to llama.cpp recently.

Thank you

youth123 committed:
Merge remote-tracking branch 'offical/master'

youth123 committed:
merge master

youth123 committed:
fix conflicts

youth123 committed:
remove prefix and suffix

youth123 committed:
use normal glm4 chattempalte & use LLM_FFN_SWIGLU in phi3

chuangfengwang commented:
Eagerly wanting this feature!

ggerganov reviewed (approved):
Let's merge after resolving conflicts

Umpire2018 committed:
fix: resolve Flake8 errors in `convert-hf-to-gguf.py`

- Fix E302 by adding two blank lines before top-level function definitions
- Replace print statements to fix NP100
- Fix E303 by ensuring only one blank line between lines of code

youth123 committed:
Merge pull request #2 from Umpire2018/fix/flake8-error

fix: resolve Flake8 errors in `convert-hf-to-gguf.py`

youth123 committed:
fix rope ratio to solve incorrect answers

youth123 committed:
merge master

youth123 commented:
I've found the cause of incorrect answers @matteoserva, @ngxson. This is because the freq_base parameter of glm4 rope is read from the config file and needed to be multiplied by 10000 when the code is actually run. After solving this problem, when I asked the model to output the first ten even numbers, the answer was correct. Then I have merged the latest code and resolved the corresponding conflicts. @ggerganov 
<img width="460" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/17959698/b2696ac8-90d0-4d37-ac85-5b373672bf7b">

ggerganov commented on convert_hf_to_gguf.py:
```diff
@@ -3176,6 +3179,190 @@ def write_tensors(self):
+        self.gguf_writer.add_file_type(self.ftype)
+        self.gguf_writer.add_rope_dimension_count(64)
+        self.gguf_writer.add_add_bos_token(False)
+        rope_ratio = 10000
+        if "rope_ratio" in self.hparams:
+            rope_ratio = rope_ratio * self.hparams["rope_ratio"]
+        self.gguf_writer.add_rope_freq_base(rope_ratio)
```
```suggestion
        rope_freq = 10000
        if "rope_ratio" in self.hparams:
            rope_freq = rope_freq * self.hparams["rope_ratio"]
        self.gguf_writer.add_rope_freq_base(rope_freq)
```

youth123 replied:
Ok, I have modified the code. When the program was running ci, a file hash consistency check failed. This does not seem to be caused by my code. Please take a look. @ggerganov 
<img width="1026" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/17959698/4523695d-98f1-4cf5-a059-36798d99d008">

youth123 committed:
fix by comments

chuangfengwang commented:
My tests show:
- glm-4-9b-chat is supported
- glm-4-9b-chat-1m is not supported

It looks like: some model layers of `glm-4-9b-chat-1m` model are different from `glm-4-9b-chat`.

Let's adapt the model `glm-4-9b-chat-1m`

youth123 commented:
> My tests show:
> 
> * glm-4-9b-chat is supported
> * glm-4-9b-chat-1m is not supported
> 
> It looks like: some model layers of `glm-4-9b-chat-1m` model are different from `glm-4-9b-chat`.
> 
> Let's adapt the model `glm-4-9b-chat-1m`
I will take a look at the model.

CsBoBoNice commented:
您好，很高兴能在该项目上用上glm-4-9b-chat模型

我使用b3333版本，使用glm-4-9b-chat模型，
发现模型一直返回不会正常停止

使用以下命令编译
```
cmake -B build -DGGML_CUDA=ON && cmake --build build --config Release -j
```

使用以下命令运行
```
./build/bin/llama-cli -m /root/autodl-tmp/glm-4-9b-chat/ggml-model-f16.gguf -ngl 41 --color -i -n 1024 -c 2048 --file ./prompts/chat-with-qwen.txt -r "User:\n" --in-suffix "Assistant:\n" --in-prefix "\n" --interactive-first
```

运行后的部分日志
```
.................................................................................
llama_new_context_with_model: n_ctx      = 2048
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 5000000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =    80.00 MiB
llama_new_context_with_model: KV self size  =   80.00 MiB, K (f16):   40.00 MiB, V (f16):   40.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     0.58 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   304.00 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    12.01 MiB
llama_new_context_with_model: graph nodes  = 1606
llama_new_context_with_model: graph splits = 2

system_info: n_threads = 64 / 128 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 1 | AVX512_VNNI = 1 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 0 |
main: interactive mode on.
Reverse prompt: 'User:
'
Input prefix: '
'
Input suffix: 'Assistant:
'
sampling:
        repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
        top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
        mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampling order:
CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature
generate: n_ctx = 2048, n_batch = 2048, n_predict = 1024, n_keep = 0


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to the AI.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.

You are a helpful assistant.
Output the first 10 even natural numbers in a row
Assistant:
2, 4, 6, 8, 10, 12, 14, 16, 18, 20. Here are the first 10 even natural numbers in a row. They are multiples of 2. Do you need any further assistance? Feel free to ask! 🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻

```

请问应该使用什么命令才能与模型正常对话?

期待您的回复

taozhiyuai commented:
https://github.com/ollama/ollama/issues/5553

chuangfengwang commented:
> 您好，很高兴能在该项目上用上glm-4-9b-chat模型
> 
> 我使用b3333版本，使用glm-4-9b-chat模型， 发现模型一直返回不会正常停止
> 
> 使用以下命令编译
> 
> ```
> cmake -B build -DGGML_CUDA=ON && cmake --build build --config Release -j
> ```
> 
> 使用以下命令运行
> 
> ```
> ./build/bin/llama-cli -m /root/autodl-tmp/glm-4-9b-chat/ggml-model-f16.gguf -ngl 41 --color -i -n 1024 -c 2048 --file ./prompts/chat-with-qwen.txt -r "User:\n" --in-suffix "Assistant:\n" --in-prefix "\n" --interactive-first
> ```
> 
> 运行后的部分日志
> 
> ```
> .................................................................................
> llama_new_context_with_model: n_ctx      = 2048
> llama_new_context_with_model: n_batch    = 2048
> llama_new_context_with_model: n_ubatch   = 512
> llama_new_context_with_model: flash_attn = 0
> llama_new_context_with_model: freq_base  = 5000000.0
> llama_new_context_with_model: freq_scale = 1
> llama_kv_cache_init:      CUDA0 KV buffer size =    80.00 MiB
> llama_new_context_with_model: KV self size  =   80.00 MiB, K (f16):   40.00 MiB, V (f16):   40.00 MiB
> llama_new_context_with_model:  CUDA_Host  output buffer size =     0.58 MiB
> llama_new_context_with_model:      CUDA0 compute buffer size =   304.00 MiB
> llama_new_context_with_model:  CUDA_Host compute buffer size =    12.01 MiB
> llama_new_context_with_model: graph nodes  = 1606
> llama_new_context_with_model: graph splits = 2
> 
> system_info: n_threads = 64 / 128 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 1 | AVX512_VNNI = 1 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 0 |
> main: interactive mode on.
> Reverse prompt: 'User:
> '
> Input prefix: '
> '
> Input suffix: 'Assistant:
> '
> sampling:
>         repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
>         top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
>         mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
> sampling order:
> CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature
> generate: n_ctx = 2048, n_batch = 2048, n_predict = 1024, n_keep = 0
> 
> 
> == Running in interactive mode. ==
>  - Press Ctrl+C to interject at any time.
>  - Press Return to return control to the AI.
>  - To return control without starting a new line, end your input with '/'.
>  - If you want to submit another line, end your input with '\'.
> 
> You are a helpful assistant.
> Output the first 10 even natural numbers in a row
> Assistant:
> 2, 4, 6, 8, 10, 12, 14, 16, 18, 20. Here are the first 10 even natural numbers in a row. They are multiples of 2. Do you need any further assistance? Feel free to ask! 🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻🤖👨‍💻🌟✨👩‍💻
> ```
> 
> 请问应该使用什么命令才能与模型正常对话?
> 
> 期待您的回复

You need special format prompt.

If you have a beginning user prompt, your command may be like this
```
./llama-cli -m path/to/glm-4-9b-chat-f16.gguf --color \
 -p "[gMASK]<|system|>\nYou are a helpful assistant<|user|>\nHello<|assistant|>\n"
 ```
 
![image](https://github.com/ggerganov/llama.cpp/assets/24692397/468159ab-5375-467a-8d10-e3a801e7a6a3)


 else your  command may be like this
```
./llama-cli -m path/to/glm-4-9b-chat-f16.gguf --color -i \
 -p "[gMASK]<|system|>\nYou are a helpful assistant<|user|>"
```
and you can interact as follow
![image](https://github.com/ggerganov/llama.cpp/assets/24692397/2db989e1-8ae2-4de8-ab29-a9b3872a04c4)

lzs0603 commented:
I've encountered an issue while setting up multiple inferencing pools with GLM4. The model continuously outputs "GGGGGGGGGGGGGGGG..." without stopping. 

The script I'm using is as follows:
```bash
#!/bin/bash
./llama-server -m /models/THUDM_glm-4-9b-chat/ggml-model-Q4_K_M.gguf -c 8192 --port 10094 --n-gpu-layers 41 -np 2 --threads 4 --host 172.17.0.1 -cb
```

When I modify `-np` to 1, the problem is resolved. Could you please help identify this issue?

hp027 commented:
same like this GGGGGG no end; 
run with cmd: ollama run glm4 
```json
{
    "title_english": "Repository Installation",
    "keyword_english": ["repository installation", "Red Hat Enterprise Linux", "CentOS", "RHEL", "Oracle Linux", "Debian", "Ubuntu", "package manager", "Zabbix", "software package"],
    "summary_english": "This document provides instructions for installing repository configuration packages on various Linux distributions to set up Zabbix server, agent, and proxy installations. It covers the process specifically for Red Hat Enterprise Linux/CentOS, Debian, and Ubuntu with different versions of supported software package managers and configurations.",
    "faq_english": [
        "What are the steps to install repository configuration packages for Zabbix on RHEL 7?",
        "How do I configure apt-get preferences for Debian installation of Zabbix repository?",
        "What versions of Linux distributions are supported for ZGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG
CPU times: user 69.9 ms, sys: 3.7 ms, total: 73.6 ms
Wall time: 7.76 s

JG-Adams commented:
I'm getting this too.
It declared, "Good Game" and gave up!

ggerganov commented:
Try similar solutions as in #8412

JG-Adams commented:
> Try similar solution as in #8412

I'm using Ollama. I don't have a clue on how to update. Does it have to wait?

tobias-varden commented:
For me the model seems fixed: `Ollama 0.2.7`

JG-Adams commented:
> For me the model seems fixed: `Ollama 0.2.7`

I have updated Ollama. It seem to be a little better I guess. But, it still have the problem.

Sakura4036 commented:
> > For me the model seems fixed: `Ollama 0.2.7`
> 
> I have updated Ollama. It seem to be a little better I guess. But, it still have the problem. I think it's the model itself that needed to be updated on Ollama



> For me the model seems fixed: `Ollama 0.2.7`

For me , `Ollama 0.2.8` not fix this problem

piDack commented:
> > > For me the model seems fixed: `Ollama 0.2.7`
> > 
> > 
> > I have updated Ollama. It seem to be a little better I guess. But, it still have the problem. I think it's the model itself that needed to be updated on Ollama
> 
> For me , `Ollama 0.2.8` not fix this problem

I tried the solution from this https://github.com/ggerganov/llama.cpp/pull/8412 and it seems to resolve the issue for me.
I’ve submitted the code. You can try the modified code on [my branch](https://github.com/piDack/llama.cpp/tree/fix_glm4_ggg_err) to see if it can resolve the issue.
